<a href="https://colab.research.google.com/github/suddeb/AppliedDataScience/blob/main/Numpy_Practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install numpy
import numpy as np

**NumPy Array**

In [4]:
# Python List
arr_str = ["Mercedes","Audi","Tesla","BMW","Ferrari"]
arr_num=[5,3,7,9,1,6]

# Converting Python List to Numpy Array
np_arr_str = np.array(arr_str)
np_arr_num = np.array(arr_num)

# Printing Numpy Array
print("Numpy Array (np_arr_str): ",np_arr_str)
print("Numpy Array (np_arr_num): ",np_arr_num)

Numpy Array (np_arr_str):  ['Mercedes' 'Audi' 'Tesla' 'BMW' 'Ferrari']
Numpy Array (np_arr_num):  [5 3 7 9 1 6]


In [5]:
# Printing Type
print("Type of Python List (arr_str): ",type(arr_str))
print("Type of Python List (arr_num): ",type(arr_num))
print("Type of Numpy Array (np_arr_str): ",type(np_arr_str))
print("Type of Numpy Array (np_arr_num): ",type(np_arr_num))

Type of Python List (arr_str):  <class 'list'>
Type of Python List (arr_num):  <class 'list'>
Type of Numpy Array (np_arr_str):  <class 'numpy.ndarray'>
Type of Numpy Array (np_arr_num):  <class 'numpy.ndarray'>


**NumPy Matrix**

In [ ]:
matrix_2d = np.array([[100,200,300],[400,500,600],[700,800,900]])
print(matrix_2d)
print("Data Type of the Matrix: ", type(matrix_2d))

[[100 200 300]
 [400 500 600]
 [700 800 900]]
Data Type of the Matrix:  <class 'numpy.ndarray'>


**NumPy Functions**

In [3]:
# List containing Integers 0 to 9
list(range(0,9)) # Python is exclusive of the last number

[0, 1, 2, 3, 4, 5, 6, 7, 8]

**NumPy Function - np.arange()**


---


This function returns an numpy array with evenly spaced elements as per the interval. The interval mentioned is half-opened i.e. start is included, but stop is excluded.
Parameters:


*   start - start of interval range. Default is 0
*   stop - end of interbal range.
*   step - step size of interval. Default is 0




In [9]:
arr = np.arange(0,10)
arr

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [7]:
# Adding a step
arr_with_step = np.arange(0,20,2)
arr_with_step

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])

**NumPy Function - np.linspace()**


---


The np.linspace() function returns numpy array of numbers which are evenly distributed with respect to interval. Here the start and stop both are included.
Parameters:


*   start - start of interval range. Default is 0
*   stop - end of interval range
*   num - number of samples to generate. Default is 50

In [10]:
my_matrix=np.linspace(0,25)
my_matrix

array([ 0.        ,  0.51020408,  1.02040816,  1.53061224,  2.04081633,
        2.55102041,  3.06122449,  3.57142857,  4.08163265,  4.59183673,
        5.10204082,  5.6122449 ,  6.12244898,  6.63265306,  7.14285714,
        7.65306122,  8.16326531,  8.67346939,  9.18367347,  9.69387755,
       10.20408163, 10.71428571, 11.2244898 , 11.73469388, 12.24489796,
       12.75510204, 13.26530612, 13.7755102 , 14.28571429, 14.79591837,
       15.30612245, 15.81632653, 16.32653061, 16.83673469, 17.34693878,
       17.85714286, 18.36734694, 18.87755102, 19.3877551 , 19.89795918,
       20.40816327, 20.91836735, 21.42857143, 21.93877551, 22.44897959,
       22.95918367, 23.46938776, 23.97959184, 24.48979592, 25.        ])

**How are these values getting calaculated?**


---
Formula = (stop - start)/(total_number_of_elements - 1)
        = (25 - 0)/49
        = 0.51020408


**Question**

---
How can we make 10 evenly spaced values between 0 and 100 ?


In [11]:
arr_1=np.linspace(0,100,10)
arr_1

array([  0.        ,  11.11111111,  22.22222222,  33.33333333,
        44.44444444,  55.55555556,  66.66666667,  77.77777778,
        88.88888889, 100.        ])

**Question**

---
How can we make evenly spaced values between 0 and 100(inclusive) where the space between two numbers will be 10?

In [13]:
arr_2=np.arange(0,101,10)
arr_2

array([  0,  10,  20,  30,  40,  50,  60,  70,  80,  90, 100])

**NumPy Function - np.zeros()**

---
This is a function for creating a matrix and performing matrix operation in NumPy. It returns a matrix filled with zeros of the given space.
Parameters:


*   shape: Number of rows and columns in the matrix
*   dtype: data type of the elements in the matrix, Default is set to float.

In [14]:
matrix_1=np.zeros([3,5])
matrix_1

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

**NumPy Function - np.ones()**


---
This is another function for creating a matrix and performing matrix operations in NumPy. It returns a matrix of given shape and type, filles with ones.
Parameter:

*   shape: Number of rows and columns in the matrix
*   dtype: data type of the elements in the matrix, Default is set to float.

In [16]:
matrix_2=np.ones([3,5])
matrix_2

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

**NumPy Function - np.eye()**

---
This is another function for creating a matrix and performating matrix operations in NumPy. It returns a matrix with ones on the diagonal and zeros elsewhere.
Parameter:

*   n: Number of rows and columns in the matrix
*   dtype: data type of the elements in the matrix, Default is set to float.





In [19]:
matrix_3=np.eye(5)
matrix_3

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

**NumPy Function - np.reshape()**

---
This function is used to convert a one dimensional array to a matrix. The shape of the array basically tells the number of the elements and dimensions of the array. Reshaping a NumPy Array simply means changing the shape of the given array. By reshaping an array, we can add or remove dimensions or change number of elements in each dimension.

In order to reshape a NumPy array, we use the reshape method with the given array.

**Syntax**: array.reshape(shape) where shape: a tuple given as input, the values in tuple will be the new shape of the array.



In [21]:
arr_3=np.arange(0,10)
arr_3

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [26]:
arr_3_reshaped=arr_3.reshape([2,5])
arr_3_reshaped

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

**Question**: What will happen when we will try to make it a 2 X 6 Matrix?

In [ ]:
arr_3_reshaped=arr_3.reshape([2,6])
arr_3_reshaped

**Explanation**: It will fail because a 2 X 6 matrix needs 12 elements and we have only 10 elements.

**NumPy Trigonometric Functions**

In [28]:
print("Sine Function: ",np.sin(4))
print("Cosine Function: ",np.cos(4))
print("Tan Function: ",np.tan(4))

Sine Function:  -0.7568024953079282
Cosine Function:  -0.6536436208636119
Tan Function:  1.1578212823495775


**NumPy Exponents Functions**

In [3]:
print("Exponent of 2: ",np.exp(2))
arr_4=np.array([2,4,6])
print("Exponent of arr_4 is: ",np.exp(arr_4))

Exponent of 2:  7.38905609893065
Exponent of arr_4 is:  [  7.3890561   54.59815003 403.42879349]


**NumPy Logarithms Functions**

In [4]:
np.log(2)

0.6931471805599453

In [5]:
np.log(arr_4)

array([0.69314718, 1.38629436, 1.79175947])

In [6]:
# log with base 10
np.log10(arr_4)

array([0.30103   , 0.60205999, 0.77815125])

**Arithmetic Operations on Arrays**

In [7]:
# Arithmetic Operations on Lists
l1=[1,2,3]
l2=[4,5,6]
print(l1+l2)

[1, 2, 3, 4, 5, 6]


In [8]:
# Arithmetic Operations on Arrays
arr_5=np.arange(1,6)
print(arr_5)

arr_6=np.arange(3,8)
print(arr_6)

[1 2 3 4 5]
[3 4 5 6 7]


In [10]:
print("Addition: ", arr_5 + arr_6)
print("Subtraction: ", arr_5 - arr_6)
print("Multiplication: ", arr_5 * arr_6)
print("Division: ", arr_5 / arr_6)
print("Inverse: ", 1 / arr_5 )
print("Powers: ", arr_5 ** arr_6)

Addition:  [ 4  6  8 10 12]
Subtraction:  [-2 -2 -2 -2 -2]
Multiplication:  [ 3  8 15 24 35]
Division:  [0.33333333 0.5        0.6        0.66666667 0.71428571]
Inverse:  [1.         0.5        0.33333333 0.25       0.2       ]
Powers:  [    1    16   243  4096 78125]


**Operations on Metrices**

In [12]:
matrix_4=np.arange(1,10).reshape(3,3)
print(matrix_4)

matrix_6=np.eye(3)
print(matrix_6)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [17]:
print("Addition: \n", matrix_4 + matrix_6)
print("Subtraction: \n", matrix_4 - matrix_6)
print("Multiplication: \n", matrix_4 * matrix_6)
print("Division: \n", matrix_4 / matrix_6)

Addition: 
 [[ 2.  2.  3.]
 [ 4.  6.  6.]
 [ 7.  8. 10.]]
Subtraction: 
 [[0. 2. 3.]
 [4. 4. 6.]
 [7. 8. 8.]]
Multiplication: 
 [[1. 0. 0.]
 [0. 5. 0.]
 [0. 0. 9.]]
Division: 
 [[ 1. inf inf]
 [inf  5. inf]
 [inf inf  9.]]


<ipython-input-17-18215cc1dabe>:4: RuntimeWarning: divide by zero encountered in divide
  print("Division: \n", matrix_4 / matrix_6)


**Linear Algebra Matrix Multiplication**

In [18]:
matrix_7=np.arange(1,10).reshape(3,3)
print("First Matrix: \n", matrix_7)

matrix_8=np.arange(11,20).reshape(3,3)
print("Second Matrix: \n", matrix_8)

print("Multiplication: \n",matrix_7 @ matrix_8)

First Matrix: 
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
Second Matrix: 
 [[11 12 13]
 [14 15 16]
 [17 18 19]]
Multiplication: 
 [[ 90  96 102]
 [216 231 246]
 [342 366 390]]


**Transpose a Matrix**

In [20]:
matrix_9=np.arange(21,30).reshape(3,3)
print("Origina: \n",matrix_9)
print("Transpose: \n", np.transpose(matrix_9)) # Another way is matrix_9.T

Origina: 
 [[21 22 23]
 [24 25 26]
 [27 28 29]]
Transpose: 
 [[21 24 27]
 [22 25 28]
 [23 26 29]]


**Find Min and Max from Matrix**

In [22]:
print("Matrix_9: \n",matrix_9)
print("Maximum Value: \n",np.max(matrix_9))
print("Minimum Value: \n",np.min(matrix_9))

Matrix_9: 
 [[21 22 23]
 [24 25 26]
 [27 28 29]]
Maximum Value: 
 29
Minimum Value: 
 21


**Using np.random.rand function**

---
The np.random.rand returns a random NumPy array whose element(s) are drawn randomly from the uniform distribution over [0,1] (Including 0 and Excluding 1)

**Synatx**: np.random.rand(d0,d1)
        d0 , d1 - It represents the dimension of the required array given as int, where d1 is optional.


In [25]:
# Random Values in the array
rand_array=np.random.rand(5)
print(rand_array)

[0.40626842 0.94907085 0.21585463 0.22214234 0.71232933]


***Using np.random.randn function***

---
This np.random.randn returns a random NumPy array whose sample(s) are drawn randomly from the standard normal ditribution (Mean as 0 and standard devioation as 1).

**Syntax**: np.random.rand(d0,d1) where do, d1 = It represents the dimention of the output where d1 is optional


In [26]:
rand_array2=np.random.randn(5)
print(rand_array2)

[ 0.62871109  1.76013516 -0.40987499  0.06586748  1.27584526]


In [29]:
rand_array3=np.random.randn(500,500)
print(rand_array3)

[[ 7.76917634e-01 -1.45642528e+00  1.36261314e+00 ...  1.10353233e+00
  -3.02795781e-01 -1.03340750e-01]
 [-9.01472817e-02  9.57986699e-01 -2.10208707e+00 ...  8.82387915e-01
   2.16226894e-01 -2.50944291e+00]
 [-1.27152535e+00 -1.25298694e+00  2.19216520e+00 ... -1.41062036e+00
  -3.11409023e-01  2.71197540e-01]
 ...
 [-1.82796126e-01 -4.39939407e-01  4.24006177e-01 ...  7.17104907e-01
   4.19784852e-01 -1.25768276e-01]
 [-1.59674556e+00 -5.47399512e-02  6.30387885e-01 ...  1.41296547e-01
  -5.49860047e-01 -7.15536376e-01]
 [ 1.36210907e+00 -1.94567341e-01  1.81192954e+00 ...  5.77598260e-01
   1.04983203e-01  3.54649437e-04]]


In [31]:
print("Mean: ",np.mean(rand_array3))
print("Standard Deviation: ",np.std(rand_array3))

Mean:  -0.002316471245687456
Standard Deviation:  0.9996440257434865


**Using np.random.randint function**

---
This np.random.randint function returns a random NumPy array whose element(s) are drwan randomly from low(inclusive) to the high(exclusive) range.

**Syntax**:  np.random.randint(low, high, size)

**Parameters**

*   low - Lowest inclusive boundary
*   high - Upper exclusive boundary
*   size - Represents the shape of the output





In [32]:
rand_mat1=np.random.randint(1,5,10)
print(rand_mat1)

[2 2 2 4 3 3 3 4 2 1]


In [33]:
rand_mat2=np.random.randint(1,10,[5,5])
print(rand_mat2)

[[2 6 5 2 7]
 [5 4 8 4 9]
 [9 3 1 6 1]
 [3 2 2 1 7]
 [4 7 2 5 2]]


**Accessing Elements of NumPy Array**

In [34]:
rand_array4=np.random.randn(10)
print(rand_array4)

[ 1.47244502 -0.17989185  0.3313149   1.50822851  1.00879256 -0.16462988
 -0.66312248  0.14970802 -0.79099911  1.152301  ]


In [35]:
# Accessing 6th Element
print(rand_array4[5])

-0.1646298805041907


In [36]:
# Accessing Multiple Elements
print(rand_array4[3:8])

[ 1.50822851  1.00879256 -0.16462988 -0.66312248  0.14970802]


In [37]:
# Accessing Non-Consecutive Elements
print("Index of values to access: ",np.arange(3, 10, 3))
print(rand_array4[np.arange(3,10,3)])


Index of values to access:  [3 6 9]
[ 1.50822851 -0.66312248  1.152301  ]


**Accessing Arrays Using Logical Operations**

In [38]:
print(rand_array4)

[ 1.47244502 -0.17989185  0.3313149   1.50822851  1.00879256 -0.16462988
 -0.66312248  0.14970802 -0.79099911  1.152301  ]


In [39]:
rand_array4>0

array([ True, False,  True,  True,  True, False, False,  True, False,
        True])

In [41]:
# Accessing only the element of the arrays whose value is greater than 0
print("Only Positive Values: ",rand_array4[rand_array4>0])
print("Only Negative Values: ",rand_array4[rand_array4<0])

Only Positive Values:  [1.47244502 0.3313149  1.50822851 1.00879256 0.14970802 1.152301  ]
Only Negative Values:  [-0.17989185 -0.16462988 -0.66312248 -0.79099911]


**Accessing the entries of a Matrix**

In [42]:
rand_mat4=np.random.randn(5,5)
print(rand_mat4)

[[ 2.52477838  0.47073486  0.0630893   0.19915571  1.00553187]
 [ 0.30963723  0.06534575 -0.31847369  0.46888536  0.40515593]
 [ 0.20487569 -0.86131782  0.2067764  -0.14588943  0.21891474]
 [ 1.16486584  0.12059225 -0.691094   -0.16874237 -0.7524963 ]
 [ 0.08698722  1.06073574  0.82176824  0.41876125 -0.68429413]]


In [43]:
# Accessing the [4,3] element from the Matrix
print(rand_mat4[4,3])

0.4187612522621179


In [45]:
# Changing the 4th & 5th element of the 2nd and 3rd rows to be 0
print("Before: \n",rand_mat4)
rand_mat4[1:3,3:5]=0
print("After: \n",rand_mat4)

Before: 
 [[ 2.52477838  0.47073486  0.0630893   0.19915571  1.00553187]
 [ 0.30963723  0.06534575 -0.31847369  0.          0.        ]
 [ 0.20487569 -0.86131782  0.2067764  -0.14588943  0.21891474]
 [ 1.16486584  0.12059225 -0.691094   -0.16874237 -0.7524963 ]
 [ 0.08698722  1.06073574  0.82176824  0.41876125 -0.68429413]]
After: 
 [[ 2.52477838  0.47073486  0.0630893   0.19915571  1.00553187]
 [ 0.30963723  0.06534575 -0.31847369  0.          0.        ]
 [ 0.20487569 -0.86131782  0.2067764   0.          0.        ]
 [ 1.16486584  0.12059225 -0.691094   -0.16874237 -0.7524963 ]
 [ 0.08698722  1.06073574  0.82176824  0.41876125 -0.68429413]]


In [46]:
# Extracting the first 2 rows and first 3 columns
sub_mat = rand_mat4[0:2,0:3]
print(sub_mat)

[[ 2.52477838  0.47073486  0.0630893 ]
 [ 0.30963723  0.06534575 -0.31847369]]


In [47]:
# Change all values to 100
sub_mat[:]=100
print(sub_mat)

[[100. 100. 100.]
 [100. 100. 100.]]


In [48]:
# What happens to rand_mat4
print(rand_mat4)

[[ 1.00000000e+02  1.00000000e+02  1.00000000e+02  1.99155709e-01
   1.00553187e+00]
 [ 1.00000000e+02  1.00000000e+02  1.00000000e+02  0.00000000e+00
   0.00000000e+00]
 [ 2.04875691e-01 -8.61317818e-01  2.06776404e-01  0.00000000e+00
   0.00000000e+00]
 [ 1.16486584e+00  1.20592251e-01 -6.91093999e-01 -1.68742373e-01
  -7.52496302e-01]
 [ 8.69872174e-02  1.06073574e+00  8.21768236e-01  4.18761252e-01
  -6.84294128e-01]]


**Reason**: Python is a reference based programming language.

In [50]:
rand_mat5=np.random.randn(5,5)
print("Original: \n",rand_mat5)
sub_mat1=rand_mat5[0:2,0:3].copy()
print("SubMat1 Before Change: \n",sub_mat1)
sub_mat1[:]=100
print("SubMat1 After Change: \n",sub_mat1)
print("Rand_Mat5: \n",rand_mat5)

Original: 
 [[ 0.60351897  0.19097058  0.95418422  0.17979464  0.63395506]
 [ 2.0038234  -1.85470902  0.89963763  0.26681219  0.48405078]
 [-0.06528036  0.49600442 -0.42245024 -1.16702407 -0.93899992]
 [ 1.03528116  1.06051035  0.6862486  -0.87708103  0.24779107]
 [ 0.88092557  1.35212363 -0.17438158  1.53652697  0.33270339]]
SubMat1 Before Change: 
 [[ 0.60351897  0.19097058  0.95418422]
 [ 2.0038234  -1.85470902  0.89963763]]
SubMat1 After Change: 
 [[100. 100. 100.]
 [100. 100. 100.]]
Rand_Mat5: 
 [[ 0.60351897  0.19097058  0.95418422  0.17979464  0.63395506]
 [ 2.0038234  -1.85470902  0.89963763  0.26681219  0.48405078]
 [-0.06528036  0.49600442 -0.42245024 -1.16702407 -0.93899992]
 [ 1.03528116  1.06051035  0.6862486  -0.87708103  0.24779107]
 [ 0.88092557  1.35212363 -0.17438158  1.53652697  0.33270339]]


**Saving and Loading a NumPy Array**

In [51]:
from google.colab import drive
drive.mount("/content/drive")

Mounted at /content/drive


In [70]:
randint_matrix1=np.random.randint(1,10,10).reshape(2,5)
print(randint_matrix1)
randint_matrix2=np.random.randint(10,20,10).reshape(2,5)
print(randint_matrix2)

[[8 8 4 8 3]
 [4 7 7 8 4]]
[[17 11 17 15 12]
 [18 16 19 16 17]]


**Using np.save() function**

In [71]:
np.save("/content/drive/MyDrive/MIT/Applied Data Science/Week 1 - Python for Data Science/file_save",randint_matrix1)

**Saving Multiple Files**

In [73]:
np.savez("/content/drive/MyDrive/MIT/Applied Data Science/Week 1 - Python for Data Science/multi_file_save",randint_matrix1=randint_matrix1,randint_matrix2=randint_matrix2,)

*   With **np.save()** function, we can save an array/matrix to a NumPy .npy format
*   With **np.savez()** function, we can store several arrays/matrices into a single file in uncompressed .npz format

In [74]:
# Let's Load
loaded_array=np.load("/content/drive/MyDrive/MIT/Applied Data Science/Week 1 - Python for Data Science/file_save.npy")
print(loaded_array)

[[8 8 4 8 3]
 [4 7 7 8 4]]


In [75]:
loaded_multi_array=np.load("/content/drive/MyDrive/MIT/Applied Data Science/Week 1 - Python for Data Science/multi_file_save.npz")
print(loaded_multi_array)
print(loaded_multi_array["randint_matrix1"])
print(loaded_multi_array["randint_matrix2"])

NpzFile '/content/drive/MyDrive/MIT/Applied Data Science/Week 1 - Python for Data Science/multi_file_save.npz' with keys: randint_matrix1, randint_matrix2
[[8 8 4 8 3]
 [4 7 7 8 4]]
[[17 11 17 15 12]
 [18 16 19 16 17]]


In [76]:
# Save as text file
np.savetxt("/content/drive/MyDrive/MIT/Applied Data Science/Week 1 - Python for Data Science/file_save.txt",randint_matrix1,delimiter=",")

In [77]:
rand_mat_txt=np.loadtxt("/content/drive/MyDrive/MIT/Applied Data Science/Week 1 - Python for Data Science/file_save.txt",delimiter=",")

print("randint_matrix1: \n",randint_matrix1)
print("rand_mat_txt: \n",rand_mat_txt)

randint_matrix1: 
 [[8 8 4 8 3]
 [4 7 7 8 4]]
rand_mat_txt: 
 [[8. 8. 4. 8. 3.]
 [4. 7. 7. 8. 4.]]
